File: java/src/main/java/com/google/oak/client/transport/GrpcStreamingTransport.java
//
// Copyright 2023 The Project Oak Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package com.google.oak.client.transport;

import static java.util.concurrent.TimeUnit.SECONDS;

import com.google.oak.client.transport.EvidenceProvider;
import com.google.oak.client.transport.Transport;
import com.google.oak.util.Result;
import com.google.protobuf.ByteString;
import io.grpc.ManagedChannel;
import io.grpc.Status;
import io.grpc.stub.StreamObserver;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.BlockingQueue;
import java.util.function.Function;
import java.util.logging.Level;
import java.util.logging.Logger;
import oak.session.noninteractive.v1.AttestationBundle;
import oak.session.noninteractive.v1.GetPublicKeyRequest;
import oak.session.noninteractive.v1.GetPublicKeyResponse;
import oak.session.noninteractive.v1.InvokeRequest;
import oak.session.noninteractive.v1.InvokeResponse;
import oak.session.noninteractive.v1.RequestWrapper;
import oak.session.noninteractive.v1.ResponseWrapper;

public class GrpcStreamingTransport implements EvidenceProvider, Transport {
  private static final Logger logger = Logger.getLogger(GrpcStreamingTransport.class.getName());

  private BlockingQueue<ResponseWrapper> messageQueue;
  private StreamObserver<RequestWrapper> requestObserver;

  /**
   * Creates an instance of {@code GrpcStreamingTransport}.
   *
   * @param stream a method reference to a gRPC client streaming method with the appropriate request
   * and response types.
   */
  public GrpcStreamingTransport(
      Function<StreamObserver<ResponseWrapper>, StreamObserver<RequestWrapper>> stream) {
    // We expect to receive a single response message for each request.
    this.messageQueue = new ArrayBlockingQueue<>(1);
    StreamObserver<ResponseWrapper> responseObserver = new StreamObserver<ResponseWrapper>() {
      @Override
      public void onNext(ResponseWrapper response) {
        try {
          logger.log(Level.INFO, "messageQueue.put(", response);
          messageQueue.put(response);
        } catch (Exception e) {
          if (e instanceof InterruptedException) {
            Thread.currentThread().interrupt();
          }
          logger.log(Level.WARNING, "Couldn't send server response to the message queue: " + e);
        }
      }

      @Override
      public void onError(Throwable t) {
        Status status = Status.fromThrowable(t);
        logger.log(Level.WARNING, "Couldn't receive response: " + status);
      }

      @Override
      public void onCompleted() {
        logger.log(Level.FINE, "response message queue completed");
      }
    };
    this.requestObserver = stream.apply(responseObserver);
  }

  /**
   * Returns evidence about the trustworthiness of a remote server.
   *
   * @return {@code AttestationBundle} wrapped in a {@code Result}
   */
  public Result<AttestationBundle, Exception> getEvidence() {
    try {
      RequestWrapper requestWrapper = RequestWrapper.newBuilder()
                                          .setGetPublicKeyRequest(GetPublicKeyRequest.newBuilder())
                                          .build();
      logger.log(Level.INFO, "sending get evidence request: " + requestWrapper);
      this.requestObserver.onNext(requestWrapper);

      ResponseWrapper responseWrapper = this.messageQueue.poll(10, SECONDS);
      logger.log(Level.INFO, "received get evidence response: " + responseWrapper);
      GetPublicKeyResponse response = responseWrapper.getGetPublicKeyResponse();

      return Result.success(response.getAttestationBundle());
    } catch (InterruptedException e) {
      return Result.error(e);
    }
  }

  /**
   * Sends a request to the enclave and returns a response.
   *
   * @param requestBytes a serialized {@code oak.crypto.EncryptedRequest} wrapped in a {@code
   *     Result}
   * @return a serialized {@code oak.crypto.EncryptedResponse} wrapped in a {@code Result}
   */
  public Result<byte[], Exception> invoke(byte[] requestBytes) {
    try {
      RequestWrapper requestWrapper =
          RequestWrapper.newBuilder()
              .setInvokeRequest(
                  InvokeRequest.newBuilder().setEncryptedBody(ByteString.copyFrom(requestBytes)))
              .build();
      logger.log(Level.INFO, "sending invoke request: " + requestWrapper);
      this.requestObserver.onNext(requestWrapper);

      ResponseWrapper responseWrapper = this.messageQueue.poll(10, SECONDS);
      logger.log(Level.INFO, "received invoke response: " + responseWrapper);
      InvokeResponse response = responseWrapper.getInvokeResponse();

      return Result.success(response.getEncryptedBody().toByteArray());
    } catch (InterruptedException e) {
      return Result.error(e);
    }
  }
}

File: cc/transport/grpc_streaming_transport.cc
/*
 * Copyright 2023 The Project Oak Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "cc/transport/grpc_streaming_transport.h"

#include <string>

#include "absl/log/log.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/string_view.h"
#include "cc/transport/util.h"
#include "grpcpp/channel.h"
#include "grpcpp/client_context.h"
#include "grpcpp/create_channel.h"
#include "grpcpp/grpcpp.h"
#include "proto/crypto/crypto.pb.h"
#include "proto/session/messages.pb.h"

namespace oak::transport {

using ::oak::crypto::v1::EncryptedRequest;
using ::oak::crypto::v1::EncryptedResponse;
using ::oak::session::v1::EndorsedEvidence;
using ::oak::session::v1::GetEndorsedEvidenceRequest;
using ::oak::session::v1::RequestWrapper;
using ::oak::session::v1::ResponseWrapper;

absl::StatusOr<EndorsedEvidence> GrpcStreamingTransport::GetEndorsedEvidence() {
  // Create request.
  RequestWrapper request;
  GetEndorsedEvidenceRequest get_endorsed_evidence_request;
  *request.mutable_get_endorsed_evidence_request() =
      get_endorsed_evidence_request;

  // Send request.
  auto response = Send(request);
  if (!response.ok()) {
    return response.status();
  }

  // Process response.
  switch (response->response_case()) {
    case ResponseWrapper::kGetEndorsedEvidenceResponseFieldNumber:
      return response->get_endorsed_evidence_response().endorsed_evidence();
    case ResponseWrapper::kInvokeResponseFieldNumber:
      return absl::InternalError(
          "received InvokeResponse instead of GetEndorsedEvidenceResponse");
    case ResponseWrapper::RESPONSE_NOT_SET:
    default:
      return absl::InternalError("received unsupported response: " +
                                 absl::StrCat(*response));
  }
}

absl::StatusOr<EncryptedResponse> GrpcStreamingTransport::Invoke(
    const EncryptedRequest& encrypted_request) {
  // Create request.
  RequestWrapper request;
  *request.mutable_invoke_request()->mutable_encrypted_request() =
      encrypted_request;

  // Send request.
  auto response = Send(request);
  if (!response.ok()) {
    return response.status();
  }

  // Process response.
  switch (response->response_case()) {
    case ResponseWrapper::kGetEndorsedEvidenceResponseFieldNumber:
      return absl::InternalError(
          "received GetEndorsedEvidenceResponse instead of InvokeResponse");
    case ResponseWrapper::kInvokeResponseFieldNumber:
      return response->invoke_response().encrypted_response();
    case ResponseWrapper::RESPONSE_NOT_SET:
    default:
      return absl::InternalError("received unsupported response: " +
                                 absl::StrCat(*response));
  }
}

GrpcStreamingTransport::~GrpcStreamingTransport() {
  absl::Status status = Close();
  if (!status.ok()) {
    LOG(WARNING) << "couldn't stop gRPC stream: " << status.message();
  }
}

absl::StatusOr<ResponseWrapper> GrpcStreamingTransport::Send(
    const RequestWrapper& request) {
  // Send a request.
  if (!channel_reader_writer_->Write(request)) {
    absl::Status status = Close();
    if (status.ok()) {
      return absl::InternalError(
          "failed to read request for unspecified reason. This is likely an "
          "implementation bug.");
    } else {
      return absl::Status(status.code(), absl::StrCat("while writing request: ",
                                                      status.message()));
    }
  }

  // Receive a response.
  ResponseWrapper response;
  if (!channel_reader_writer_->Read(&response)) {
    absl::Status status = Close();
    if (status.ok()) {
      return absl::InternalError(
          "failed to write request for unspecified reason. This is likely an "
          "implementation bug.");
    } else {
      return absl::Status(status.code(), absl::StrCat("while reading request: ",
                                                      status.message()));
    }
  }
  return response;
}

absl::Status GrpcStreamingTransport::Close() {
  absl::call_once(close_once_, [this]() {
    channel_reader_writer_->WritesDone();
    grpc::Status grpc_close_status = channel_reader_writer_->Finish();
    close_status_ = to_absl_status(grpc_close_status);
  });
  return close_status_;
}

}  // namespace oak::transport
